DSharpPlus.Commands/ContextChecks/ParameterChecks/RequireHierarchyCheck.cs

```diff
@@ -0,0 +1,100 @@
+/// </summary>
+public sealed class RequireHierarchyCheck : 
+    IParameterCheck<RequireHigherBotHierarchyAttribute>,
+    IParameterCheck<RequireHigherUserHierarchyAttribute>
+{
+    public ValueTask<string?> ExecuteCheckAsync
+    (
+        RequireHigherBotHierarchyAttribute attribute, 
+        ParameterCheckInfo info, 
+        CommandContext context
+    )
+    {
+        // no value is always hierarchically lower than any user
+        if (info.Value is null)
+        {
+            return ValueTask.FromResult<string?>(null);
+        }
+
+        // no guild means no hierarchy.
+        if (context.Guild is null)
+        {
+            return ValueTask.FromResult<string?>(null);
+        }
+
+        if (info.Value is DiscordRole role)
+        {
+            return ValueTask.FromResult
+            (
+                context.Guild.CurrentMember.Hierarchy > role.Position
+                    ? null
+                    : "The provided role was higher than the highest role of the bot user."
+            );
+        }
+
+        if (info.Value is DiscordMember member)
+        {
+            return ValueTask.FromResult
+            (
+                context.Guild.CurrentMember.Hierarchy > member.Hierarchy
+                    ? null
+                    : "The provided member's highest role was higher than the highest role of the bot user."
+            );
+        }
+
+        return ValueTask.FromResult<string?>("The provided parameter was neither a role nor an user, failed to check hierarchy.");
+    }
+
+    public ValueTask<string?> ExecuteCheckAsync
+    (
+        RequireHigherUserHierarchyAttribute attribute, 
+        ParameterCheckInfo info,
+        CommandContext context
+    )
+    {
+        // no value is always hierarchically lower than any user
+        if (info.Value is null)
+        {
+            return ValueTask.FromResult<string?>(null);
+        }
+
+        // no guild means no hierarchy.
+        if (context.Guild is null)
+        {
+            return ValueTask.FromResult<string?>(null);
+        }
+
+        if (info.Value is DiscordRole role)
+        {
+            return ValueTask.FromResult
+            (
+                // this should? never be null if we're in a guild
+                context.Member!.Hierarchy > role.Position
+                    ? null
+                    : "The provided role was higher than the highest role of the executing user."
+            );
+        }
+
+        if (info.Value is DiscordMember member)
+        {
+            return ValueTask.FromResult
+            (
+                context.Member!.Hierarchy > member.Hierarchy
+                    ? null
+                    : "The provided member's highest role was higher than the highest role of the executing user."
+            );
+        }
+
+        return ValueTask.FromResult<string?>("The provided parameter was neither a role nor an user, failed to check hierarchy.");
+    }
```

Comment:
```suggestion
    public ValueTask<string?> ExecuteCheckAsync(RequireHigherUserHierarchyAttribute attribute, ParameterCheckInfo info, CommandContext context) => info.Value switch
    {
        _ when context.Guild is null => ValueTask.FromResult<string?>(null),
        DiscordRole role when context.Member!.Hierarchy > role.Position => ValueTask.FromResult<string?>(null),
        DiscordRole => ValueTask.FromResult("The provided role was higher than the highest role of the executing user."),
        DiscordMember member when context.Member!.Hierarchy > member.Hierarchy => ValueTask.FromResult<string?>(null),
        DiscordMember => ValueTask.FromResult("The provided member's highest role was higher than the highest role of the executing user."),
        _ => ValueTask.FromResult<string?>("The provided parameter was neither a role nor an user, failed to check hierarchy.")
    };
```
:))

Reply:
don't know how well i like the readability of this, but fine